### PR TITLE
Support vpiInertialDelay

### PR DESCRIPTION
**Pull request by toddstrader**
Adds `vpiInertialDelay` support for `vpi_put_value()`.

### CODE REVIEW

File: include/verilated_vpi.cpp [hunk 1]
```diff
@@ -2447,6 +2550,17 @@ vpiHandle vpi_put_value(vpiHandle object, p_vpi_value valuep, p_vpi_time /*time_
         return nullptr;
     }
     if (const VerilatedVpioVar* const vop = VerilatedVpioVar::castp(object)) {
+        if (flags & vpiInertialDelay) {
```
Comment (wsnyder) [inline include/verilated_vpi.cpp:2553]: My reading of IEEE is this isn't a bit mask but absolute number.  (If not sure look at Icarus?)

```suggestion
        if (flags == vpiInertialDelay) {
```

File: include/verilated_vpi.cpp [hunk 2]
```diff
@@ -2447,6 +2550,17 @@ vpiHandle vpi_put_value(vpiHandle object, p_vpi_value valuep, p_vpi_time /*time_
...
+            if (!VerilatedVpiPutHolder::canInertialDelay(valuep)) {
+                VL_VPI_ERROR_(
+                    __FILE__, __LINE__,
+                    "%s: Unsupported p_vpi_value as requested for %s with vpiInertialDelay",
```
Comment (wsnyder) [inline include/verilated_vpi.cpp:2557]:
```suggestion
                    "%s: Unsupported p_vpi_value as requested for '%s' with vpiInertialDelay",
```

Also will need a test to cover this case.

File: include/verilated_vpi.cpp [hunk 3]
```diff
@@ -630,6 +630,96 @@ class VerilatedVpiCbHolder final {
...
+        case vpiHexStrVal:
+        case vpiStringVal: {
+            if (VL_UNLIKELY(!valuep->value.str)) { return false; }
+            break;
+        }
+        case vpiScalarVal:
+        case vpiIntVal:
+        case vpiRealVal: break;
+        case vpiVectorVal: {
+            if (VL_UNLIKELY(!valuep->value.vector)) { return false; }
```
Comment (wsnyder) [inline include/verilated_vpi.cpp:712]:
```suggestion
            if (VL_UNLIKELY(!valuep->value.str)) return false;
            break;
        }
        case vpiScalarVal:
        case vpiIntVal:
        case vpiRealVal: break;
        case vpiVectorVal: {
            if (VL_UNLIKELY(!valuep->value.vector)) return false;
```

File: include/verilated_vpi.cpp [hunk 4]
```diff
@@ -630,6 +630,96 @@ class VerilatedVpiCbHolder final {
...
+        switch (valuep->format) {
+        case vpiBinStrVal:
+        case vpiOctStrVal:
+        case vpiDecStrVal:
+        case vpiHexStrVal:
+        case vpiStringVal: {
+            m_str = valuep->value.str;
+            m_value.value.str = &m_str[0];
+            break;
+        }
```
Comment (wsnyder) [inline include/verilated_vpi.cpp:654]: Is this function duplicated code we should put into a subroutine?

**Review by wsnyder — COMMENTED** [review]
Is this compatible with how other simulators handle vpiInertialDelay?  Did you try the test on another simulator?

File: include/verilated_vpi.cpp (hunk 1 above)
Comment (toddstrader) [reply include/verilated_vpi.cpp:2553]: > My reading of IEEE is this isn't a bit mask but absolute number. (If not sure look at Icarus?)

Super weird.  `flags` is both an enum value and a bit vector combined.  I think the only bit vector field that applies to `vpi_put_value()` is `vpiReturnEvent`.  I don't see a formal definition of what to mask off when comparing the enum portion of `flags`, but from looking at `vpi_user.h`, I think it is probably `0xfff`.

I'm not sure I agree with the LRM's use of the term `flag` here, but oh well.

File: include/verilated_vpi.cpp (hunk 4 above)
Comment (toddstrader) [reply include/verilated_vpi.cpp:654]: > Is this function duplicated code we should put into a subroutine?

I don't believe so.  It's setting some members of a class which was created for this PR, so I don't believe it could have existed beforehand.  And there's nothing else that sets `m_str`.

But maybe I'm missing your point.  If there's something else that does this can you point me at it so I can refactor?

**Comment by toddstrader** [comment]
> Is this compatible with how other simulators handle vpiInertialDelay? Did you try the test on another simulator?

I tried to get `t_vpi_var` working with other simulators for a bit (icarus, xcelium, xsim) but everything has significant problems with that test.

I believe this more or less demonstrates `vpiInertialDelay` with other simulators:
https://edaplayground.com/x/aDri

Basically if you `vpi_put_value()` with `vpiInertialDelay` and then immediately `vpi_get_value()` you will not see your new value.  But once you return control to the simulator the new value is applied.

Of course, Verilator handles all the VPI concerns outside of `eval()`, so when to call `doInertialPuts()` isn't really a Verilator-proper concern but instead would be left to the users.

One other thing wrt other simulators, I think Verilator is handling the return value for `vpi_put_value()` incorrectly:
```
If the flags argument also has the bit mask vpiReturnEvent, vpi_put_value() shall return a handle of type
vpiSchedEvent to the newly scheduled event, provided there is some form of a delay and an event is
scheduled. If the bit mask is not used, or if no delay is used, or if an event is not scheduled, the return value
shall be NULL .
```

Currently we return the object handle if things are good and `nullptr` if not.  However, I don't think we should ever return the object handle.  But I'm not currently concerned with the return value here or canceling scheduled events, so I'm sticking with the status quo of returning the object handle on success.

File: include/verilated_vpi.cpp (hunk 2 above)
Comment (toddstrader) [reply include/verilated_vpi.cpp]: > Also will need a test to cover this case.

Do you have any guidance for what should be a `VL_VPI_WARNING_` vs `VL_VPI_ERROR_`?  I changed this to a warning which made adding checks to `t_vpi_var` easy since it doesn't cause the process to abort.  However, now I'm not sure which way is correct.  E.g. supplying an unsupported format is an error, but an unsupported object type is a warning.

Also, if I go back to an error, I believe I'll need to make a bunch of `_bad` tests (of which there are currently none for VPI as far as I can tell) since each incorrect call will kill the process.

**Comment by toddstrader** [comment]
I think I've covered all of the comments with the latest commit.  Also there are some new questions from me sprinkled in above.

File: include/verilated_vpi.cpp (hunk 2 above)
Comment (wsnyder) [reply include/verilated_vpi.cpp]: I generally use warning, as it seems the VPI was assuming that in most cases, so I'm good with your choice.
